Add netconf_options to allow passing args to netconf_config

oponcea-dn opened the pull request:
##### SUMMARY
Netconf `ansible.netcommon.netconf_config` has many options that makes it flexible. So, although `community.yang.configure` provides good functionality on top of `netconf`, it can't be used at full strength as we may need features of `netconf_config` that are not accessible.

Given that `community.yang.configure` is built on top of `ansible.netcommon.netconf_config`, we can easily pass argument options directly to it. Therefore, this PR adds a new option, `netconf_options`, to allow users to pass options directly to` neconf_config`.

The DOCUMENTION for the suboptions of `neconf_options` are taken from `netconf_config` with the following exceptions:
- `content, source_datastore, delete, confirm_commit` are removed; `content` is always set by `community.yang.configure` and  they are mutually exclusive.
- `format` is removed: this is the format of data passed as input in content field an supported by the device, yang module prepares the data for the content option in a single format so no need to support this option
- `get_filter` is removed: already set by get_filter of `community.yang.configure`.

##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
community.yang.configure

##### ADDITIONAL INFORMATION
Usage:
```yaml
- name: pass arguments to netconf_config
  community.yang.configure:
    config: "{{ lookup('file', 'interfaces-config.json') }}"
    file: "{{ playbook_dir }}/public/release/models/interfaces/openconfig-interfaces.yang"
    search_path: "{{ playbook_dir }}/public/release/models"
    netconf_options:
      lock: never
      username: system
      password: complex_password
```

Error when setting a netconf_config invalid option:
```
TASK [Execute yang netconf_config] **************************************************************************************************************
task path: /home/dn/workspace/ansible/yang-features/demo_04_yang_configure.yml:6
fatal: [dn-hw-ncc]: FAILED! => {
    "changed": false,
    "msg": "Unsupported parameters for (basic.pyc) module: aaaa found in netconf_options. Supported parameters include: backup, backup_options, commit, confirm, confirm_commit, default_operation, delete, error_option, lock, save, source_datastore, target, validate"
}
```

oponcea-dn commented:
Zuul sanity fails with:
```
2021-05-05 08:57:23.595242 | centos-8 | Run command: ansible-doc -t lookup community.yang.json2xml community.yang.spec community.yang.xml2json
2021-05-05 08:57:24.247249 | centos-8 | ERROR: Output on stderr from ansible-doc is considered an error.
2021-05-05 08:57:24.247329 | centos-8 |
2021-05-05 08:57:24.247341 | centos-8 | Command "ansible-doc -t lookup community.yang.json2xml community.yang.spec community.yang.xml2json" returned exit status 0.
2021-05-05 08:57:24.247348 | centos-8 | >>> Standard Error
2021-05-05 08:57:24.247354 | centos-8 | [WARNING]: Collection community.yang does not support Ansible version
2021-05-05 08:57:24.247371 | centos-8 | 2.12.0.dev0[0m
```
Which does not seem to be related to changes in this PR as `source ~/venv/bin/activate; ansible-test sanity --requirements --docker -v` passed locally, on my machine.

oponcea-dn commented:
@ganeshrn , how does this change looks to you? I can't add reviewers so, other than writing you directly, I don't see many options to get in touch :)

Thanks!

PS I just joined the IRC channel as ovidiuponcea, I can't join the slack one as I need an invite

ganeshrn reviewed (commented):
@oponcea-dn Thanks for your PR and contribution. The functionality added is certianly helpful, however I think if we add the `netconf_config` options which are not deprecated in documentation as part of `netconf_config_args` as suboptions [example](https://github.com/ansible-collections/ansible.netcommon/blob/main/plugins/modules/netconf_config.py#L174), it will be more clear from user point of view. The `content` and `get_filter` options can  referred from `communit.yang.configure` module itself.

ganeshrn commented:
To add to above comment `netconf_config` module is pretty stable and no new options are added to that module for quite some time now. Also we do have plans to add support for restconf transport as well so the `netconf_config_args` and `restconf_config_args` options can be multually exclusive when they are added. What do you think?

ganeshrn commented:
recheck

oponcea-dn commented:
Yeah, netconf_config is stable and it makes sense to filter those that are deprecated (although none is currently in this state but may become in the future) & to also have all of them listed in a single doc page instead of instructing users to go and grab them from another module; yet I avoided it for three reasons:
1. netconf_config may change and it will need a commit with each new option added/updated;
2. I did not want to copy paste the description from the other component (same reason, it can change and it will be hard to notice)
3. the suboptions are validated automatically when netconf_config is called, so in case we use an option that is not supported, it will be rejected by netconf_config when yang.configure calls it.

If the above reasons are acceptable, if it's ok with you and given the stability of netconf_config, I'll make the adjustments: add them as suboptions by taking the docstrings from netconf_config.

(and great news regarding restconf :) )

ashwini-mhatre commented:
recheck

ganeshrn commented:
@oponcea-dn Your concerns on `netconf_config` module getting change is valid but with the current approach, the `content` and `get-filter` options are duplicated and it has scope for confusion. IMO it is more user friendly to have the docs of valid and non-deprecated options of `netconf_config`  module as part of `community.yang.configure` module docs within suboptions (say `netconf_options` or `netconf_args` instead of `netconf_config_args`).
The user of `community.yang.configure` doesn't need to know that the module uses `netconf_config` under the hood.

In addition to that, we can add a note in `netconf_config` module to update the `community.yang.configure` doc whenever it is changed. I know it is not an optimised solution but considering the frequency of update in `netconf_config` module, I don't see it as much of an issue.

oponcea-dn commented:
Super, thanks! Then I'll proceed with the changes:
1. Rename` netconf_config_args` to `neconf_options`: although both `args` and `options` are used interchangeably throughout the code, I just saw `options` is used much more throughout Ansible yamls while `args` throughout modules python code.
2. Copy the options from netconf_config to yang.configure

ganeshrn commented:
> Super, thanks! Then I'll proceed with the changes:
> 
> 1. Rename` netconf_config_args` to `neconf_options`: although both `args` and `options` are used interchangeably throughout the code, I just saw `options` is used much more throughout Ansible yamls while `args` throughout modules python code.
> 2. Copy the options from netconf_config to yang.configure

+1 Thanks!

ganeshrn commented:
Also please do add integration tests if feasible. We currently test against iosxr and junos in the CI refer https://github.com/ansible-collections/community.yang/tree/main/tests/integration/targets/configure/tasks

oponcea-dn commented:
Updated:
1. Renamed `netconf_config_args` to `neconf_options`
2. Copied the DOCUMENTATION options from `netconf_config` to `netconf_options`
3. Removed the following `netconf_config` options that no longer make sense in `community.yang.configure`:
- since content is always set by `community.yang.configure` and these netconf parameters are mutually exclusive: content|source_datastore|delete|confirm_commit all of them are removed.
- `format`: this is the format of data passed as input in content field an supported by the device, yang module prepares the data for the content option in a single format so no need to support this option
- `get_filter`: already set by `get_filter` of `community.yang.configure`.

oponcea-dn commented:
> Also please do add integration tests if feasible. We currently test against iosxr and junos in the CI refer https://github.com/ansible-collections/community.yang/tree/main/tests/integration/targets/configure/tasks

Since we do not have Cisco devices we can't do much about it :). Thanks!

ganeshrn reviewed (approved):
lgtm, Thanks for your contribution.

ganeshrn commented:
> > Also please do add integration tests if feasible. We currently test against iosxr and junos in the CI refer https://github.com/ansible-collections/community.yang/tree/main/tests/integration/targets/configure/tasks
> 
> Since we do not have Cisco devices we can't do much about it :). Thanks!

Have raised an issue to track it https://github.com/ansible-collections/community.yang/pull/58#issuecomment-838430991

ganeshrn commented:
recheck

oponcea-dn commented:
recheck

ganeshrn commented:
recheck
